src/x/server/secured_conn.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package server implements a network server.
+package server
+
+import (
+	"crypto/tls"
+	"net"
+	"sync"
+)
+
+// TLSHandshakeFirstByte is the first byte of a tls connection handshake
+const TLSHandshakeFirstByte = 0x16
+
+func newSecuredConn(conn net.Conn) *securedConn {
+	return &securedConn{
+		Conn: conn,
+	}
+}
+
+type securedConn struct {
+	net.Conn
+	mu         sync.Mutex
+	isTLS      *bool
+	peekedByte *byte
+}
+
+// IsTLS returns is the connection is TLS or not.
+// It peeks at the first byte and checks
+// if it is equal to the TLS handshake first byte
+// https://www.rfc-editor.org/rfc/rfc5246#appendix-A.1
+func (s *securedConn) IsTLS() (bool, error) {
+	s.mu.Lock()
```

Comment:
Not sure I understand your locking here (was this here before?). Is this ever called in a multithreaded context? (iirc, readers in general are not required to be thread safe).